### PR TITLE
coprocessor, dag: add like signature.

### PR DESCRIPTION
**Pull request by hicqu**

### CODE REVIEW

File: src/coprocessor/dag/expr/compare.rs [hunk 1]
```diff
@@ -157,6 +158,20 @@ impl FnCall {
     ) -> Result<Option<Cow<'a, Json>>> {
         do_coalesce(self, |v| v.eval_json(ctx, row))
     }
+
+    pub fn like(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let target = try_opt!(self.children[0].eval_string(ctx, row));
+        let target = try!(str::from_utf8(&target).map_err(Error::from));
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/compare.rs]: What if it's not a valid unicode string?

Comment (hicqu) [reply src/coprocessor/dag/expr/compare.rs]: Here I followed TiDB.
`children[0]` is autowrapped with `cast as string` before step in here, so I think if it's a invalid unicode string, the error has already been checked and return in `cast`?

Comment (BusyJay) [reply src/coprocessor/dag/expr/compare.rs]: I don't think so. TiDB supports latin1 charset, which is not unicode compatible.

Comment (hicqu) [reply src/coprocessor/dag/expr/compare.rs]: got it. Find [something](https://stackoverflow.com/questions/28169745/what-are-the-options-to-convert-iso-8859-1-latin-1-to-a-string-utf-8/28175593#28175593) to fix this.

Comment (ngaut) [reply src/coprocessor/dag/expr/compare.rs]: Does TiDB support latin1 charset ? @shenli

File: src/coprocessor/dag/expr/compare.rs [hunk 2]
```diff
@@ -230,15 +245,50 @@ where
...
+                    if t == p && like_match(tcs.as_str(), pcs.as_str(), escape) {
+                        return true;
+                    }
+                }
+                return false;
+            }
+            return true;
+        } else {
+            if let Some(t) = tcs.next() {
+                if t == c || c == '_' {
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/compare.rs]: Is like case-sensitive?

Comment (hicqu) [reply src/coprocessor/dag/expr/compare.rs]: By default, it's case-insensitive.

File: src/coprocessor/dag/expr/compare.rs [hunk 3]
```diff
@@ -157,6 +158,20 @@ impl FnCall {
     ) -> Result<Option<Cow<'a, Json>>> {
         do_coalesce(self, |v| v.eval_json(ctx, row))
     }
+
+    pub fn like(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let target = try_opt!(self.children[0].eval_string(ctx, row));
+        let target = try!(str::from_utf8(&target).map_err(Error::from));
+        let pattern = try_opt!(self.children[1].eval_string(ctx, row));
+        let pattern = try!(str::from_utf8(&pattern).map_err(Error::from));
+        let escape = if self.children.len() == 3 {
```
Comment (hanfei1991) [inline src/coprocessor/dag/expr/compare.rs]: You can suppose the length of children **must be** 3

File: src/coprocessor/dag/expr/compare.rs (hunk 1 above)
Comment (hicqu) [reply src/coprocessor/dag/expr/compare.rs]: @ngaut, we only support UTF-8. Here I will rewrite a new function `eval_string_and_decode` which returns `Cow<str>`, and do string decode logic in that.

**Comment by hicqu** [comment]
@BusyJay @AndreMouche PTAL again, thanks.

**Comment by hicqu** [comment]
@XuHuaiyu , both TiKV and TiDB should follow MySQL's behaviour. Please make `LIKE` in TiDB case-insensitive.

**Comment by XuHuaiyu** [comment]
@hicqu got it.

**Comment by hanfei1991** [comment]
@hicqu If we keep consistent with MySQL, the `like` predicate cannot use index any more.

**Comment by hicqu** [comment]
So, after discuss about this, for now we are going to keep `LIKE` case-sensitive. @XuHuaiyu @BusyJay .

**Comment by hicqu** [comment]
@BusyJay @AndreMouche PTAL, thanks.

File: src/coprocessor/dag/expr/compare.rs [hunk 4]
```diff
@@ -157,6 +158,16 @@ impl FnCall {
     ) -> Result<Option<Cow<'a, Json>>> {
         do_coalesce(self, |v| v.eval_json(ctx, row))
     }
+
+    pub fn like(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let target = try_opt!(self.children[0].eval_string_and_decode(ctx, row));
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/compare.rs]: Does target have to be a string? Can it be a blob?

Comment (hicqu) [reply src/coprocessor/dag/expr/compare.rs]: If call LIKE on `blob`, TiDB type infer mechanism will convert that to `String`. Now I don't know how mysql convets `blob` to `string`. Both TiDB and TiKV treats those bytes as **UTF8**. I will figure that.

Comment (hicqu) [reply src/coprocessor/dag/expr/compare.rs]: If we can't treat blob bytes as UTF8, I think we should support the real codec in `eval_string_and_decode` and call that function in `cast_string_as_string`. Maybe raise a new PR is better?

File: src/coprocessor/codec/mysql/charset.rs [hunk 5]
```diff
@@ -31,3 +31,6 @@ pub const COLLATION_ASCII: &'static str = "ascii_bin";
 pub const CHARSET_LATIN1: &'static str = "latin1";
 // `COLLATION_LATIN1` is the default collaction for `CHARSET_LATIN1`.
 pub const COLLATION_LATIN1: &'static str = "latin1_bin";
+
+// All utf8 charsets.
+pub const UTF8_CHARSETS: [&'static str; 3] = [CHARSET_UTF8, CHARSET_UTF8MB4, CHARSET_ASCII];
```
Comment (BusyJay) [inline src/coprocessor/codec/mysql/charset.rs]: Use `&'static [&'static str]` instead.

File: src/coprocessor/dag/expr/fncall.rs [hunk 6]
```diff
@@ -206,7 +206,9 @@ impl FnCall {
 
             ScalarFuncSig::JsonExtractSig |
             ScalarFuncSig::JsonRemoveSig |
-            ScalarFuncSig::JsonMergeSig => (2, usize::MAX),
+            ScalarFuncSig::JsonMergeSig |
+            ScalarFuncSig::LikeSig => (2, usize::MAX),
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/fncall.rs]: I think it should be (3, 3).

File: src/coprocessor/dag/expr/mod.rs [hunk 7]
```diff
@@ -205,6 +205,23 @@ impl Expression {
...
     }
 
+    fn eval_string_and_decode<'a, 'b: 'a>(
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, str>>> {
+        let bytes = try_opt!(self.eval_string(ctx, row));
+        let chrst = self.get_tp().get_charset();
+        if charset::UTF8_CHARSETS.iter().any(|&c| c == chrst) {
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/mod.rs]: Use `contains` instead.

File: src/coprocessor/dag/expr/compare.rs [hunk 8]
```diff
@@ -230,15 +241,51 @@ where
     Ok(None)
 }
 
+#[inline]
+fn next_escaped(chars: &mut Chars, escape: char) -> Option<char> {
+    chars.next().and_then(|c| if c == escape {
+        chars.next().or_else(|| Some(c))
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/compare.rs]: I think this is wrong. "\%" should match "%" exactly instead of any string.

Comment (hicqu) [reply src/coprocessor/dag/expr/compare.rs]:
```sql
mysql> select 'hello' like '%';
+------------------+
| 'hello' like '%' |
+------------------+
|                1 |
+------------------+
1 row in set (0.00 sec)
```
It comes from MySQL.

Comment (BusyJay) [reply src/coprocessor/dag/expr/compare.rs]: Github ate my "\\", what I mean is "\\%".

**Comment by hicqu** [comment]
@BusyJay , PTAL thanks.

File: src/coprocessor/dag/expr/compare.rs [hunk 9]
```diff
@@ -342,4 +391,42 @@ mod test {
...
+            (r#"C:\Programs\"#, r#"%\\"#, '\\', true),
+            (r#"C:\Programs\"#, r#"%Prog%"#, '\\', true),
+            (r#"C:\Programs\"#, r#"%Pr_g%"#, '\\', true),
+            (r#"C:\Programs\"#, r#"%%\"#, '%', true),
+            (r#"C:\Programs%"#, r#"%%%"#, '%', true),
+            (r#"C:\Programs%"#, r#"%%%%"#, '%', true),
+            (r#"hello"#, r#"\%"#, '\\', false),
+            (r#"%"#, r#"\%"#, '\\', true),
+            (r#"3hello"#, r#"%%hello"#, '%', true),
+            (r#"3hello"#, r#"__hello"#, '_', true),
```
[the remark this answers is not in the document]
Comment (hicqu) [reply src/coprocessor/dag/expr/compare.rs]: It's a strange behavior. But I can follow it.

[the remark this answers is not in the document]
Comment (BusyJay) [reply src/coprocessor/dag/expr/compare.rs]: I think cases like `"3hello" like "%%hello" escape '%'` is the strange one actually.

File: src/coprocessor/dag/expr/compare.rs [hunk 10]
```diff
@@ -230,15 +241,48 @@ where
     Ok(None)
 }
 
+#[inline]
+fn next_escaped(chars: &mut Chars, escape: char) -> Option<(char, bool)> {
+    chars.next().map(|c| if c == escape {
+        chars.next().map(|c| (c, true)).unwrap_or((c, false))
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/compare.rs]: Use `map_or` instead.

File: src/coprocessor/dag/expr/mod.rs [hunk 11]
```diff
@@ -205,6 +205,23 @@ impl Expression {
         }
     }
 
+    fn eval_string_and_decode<'a, 'b: 'a>(
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, str>>> {
+        let bytes = try_opt!(self.eval_string(ctx, row));
+        let chrst = self.get_tp().get_charset();
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/mod.rs]: Is the charset always set?

Comment (hicqu) [reply src/coprocessor/dag/expr/mod.rs]: I think so, TiDB will set it to `UTF8` by default. See `tidb/expression/builtin_cast.go:1342`.

File: src/coprocessor/dag/expr/compare.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (hicqu) [reply src/coprocessor/dag/expr/compare.rs]: yes, `3hello LIKE "%_hello" escape '%'` also returns `true`.

**Comment by hicqu** [comment]
We don't have full collate support. the behavior of `LIKE` depends on that. So I think for now keeping it case-sensitive is OK, just like we keep `'abc' != 'ABC'`. :)

**Comment by BusyJay** [comment]
LGTM

**Pull request merged**